tests/karma/unit/services/contact-save.js

```diff
@@ -0,0 +1,63 @@
+var assert = chai.assert;
+
+describe('ContactSave service', function() {
+
+  var service;
+  var EnketoTranslation;
+
+  var bulkDocsHandler;
+  var contactRecordsJs;
+
+  beforeEach(function() {
+    EnketoTranslation = {
+      contactRecordToJs: function() {
+        return contactRecordsJs;
+      },
+    };
+
+    module('inboxApp');
+    module(function($provide) {
+      $provide.value('$q', Q); // bypass $q so we don't have to digest
+      $provide.value('DB', function() { return { bulkDocs: bulkDocsHandler }; });
+      $provide.value('EnketoTranslation', EnketoTranslation);
+      $provide.value('ExtractLineage', function(c) { return c; });
+    });
+    inject(function(_ContactSave_) {
+      service = _ContactSave_;
+    });
+  });
+
+  it('should include parent ID in repeated children', function() {
+
+    var schema = { fields: { sis: { type: 'db:sister', }, }, };
+    var form = { getDataStr: function() { return '<data></data>'; } };
+    var docId = null;
+    var type = 'some-contact-type';
+
+    contactRecordsJs = {
+      doc: { _id: 'main1', type: 'main', sis: 'NEW', },
+      siblings: {
+        sis: { _id: 'sis1', type: 'sister', parent: 'PARENT', },
+      },
+      repeats: {
+        child_data: [ { _id: 'kid1', type: 'child', parent: 'PARENT', } ],
+      },
+    };
+
+    bulkDocsHandler = function(docs) {
+
+      assert.equal(docs[0]._id, 'kid1');
+      assert.equal(docs[0].parent._id, 'main1');
+
+      assert.equal(docs[1]._id, 'sis1');
+      assert.equal(docs[1].parent._id, 'main1');
+
+      assert.equal(docs[2]._id, 'main1');
+
+      return Promise.resolve([]);
+    };
+
+    return service(schema, form, docId, type);
+  });
+
+});
```

Comment:
Loving the test! Nice one.

Can you convert it to ES6 styles? Closures, const/let, etc.

I find sinon easier to follow than having your own bulkDocsHandler mock.

It's probably also worth making some assertions about what the service returns.

Reply:
Also worth checking that ExtractLineage is called?

Reply:
> Can you convert it to ES6 styles? Closures, const/let, etc.

I didn't do this because it's a karma test, so presumably runs in the browser.  I guess it's not an issue, but it's a shame we don't run these tests on Chrome 30.

Reply:
> I find sinon easier to follow than having your own bulkDocsHandler mock.

Would `DB` be a sinon object, or would `bulkDocsHandler`?

Reply:
> I didn't do this because it's a karma test, so presumably runs in the browser.

Yeah. We're using it in other places, but that's a fair call.

> Would DB be a sinon object, or would bulkDocsHandler?

In other tests we do something like...

```js
bulkDocs = sinon.stub();
$provide.factory('DB', KarmaUtils.mockDB({ bulkDocs: bulkDocs }));
```

Reply:
This isn't a drop-in replacement.  Doing this gives:
```
1) should include parent ID in repeated children
     ContactSave service
     TypeError: e(...).bulkDocs is not a function
    at static/dist/inbox.js:3:2095867
    at _fulfilled (tests/karma/q.js:834:54)
    at self.promiseDispatch.done (tests/karma/q.js:863:30)
    at Promise.promise.promiseDispatch (tests/karma/q.js:796:13)
    at tests/karma/q.js:604:44
    at runSingle (tests/karma/q.js:137:13)
    at MessagePort.flush (tests/karma/q.js:125:13)
```